example/src/main/java/io/netty/example/tunechoserver/InetUtils.java

```diff
@@ -0,0 +1,399 @@
+
+        return new InetUtils.UDPHeader(sourcePort, destPort, length, checksum);
+    }
+
+    public static final class TCPHeader {
+        public final int sourcePort;
+        public final int destPort;
+        public final long sequenceNum;
+        public final int flags;
+        public final int dataLength;
+        public final int checksum;
+
+        public TCPHeader(int sourcePort, int destPort, long sequenceNum, int flags, int dataLength, int checksum) {
+            this.sourcePort = sourcePort;
+            this.destPort = destPort;
+            this.sequenceNum = sequenceNum;
+            this.flags = flags;
+            this.dataLength = dataLength;
+            this.checksum = checksum;
+        }
+    }
+
+    public static TCPHeader decodeTCPHeader(ByteBuf packetBuf, int headerOffset) {
+        // Fail if the packet is too small.
+        if (packetBuf.readableBytes() < headerOffset + TCP_HEADER_LENGTH) {
+            throw new IllegalArgumentException("TCP packet too small");
+        }
+
+        int sourcePort = packetBuf.getUnsignedShort(headerOffset + TCP_HEADER_OFFSET_SOURCE_PORT);
+        int destPort = packetBuf.getUnsignedShort(headerOffset + TCP_HEADER_OFFSET_DEST_PORT);
+        long sequenceNum = packetBuf.getUnsignedInt(headerOffset + TCP_HEADER_OFFSET_SEQUENCE_NUM);
+        int dataOffsetAndFlags = packetBuf.getUnsignedShort(headerOffset + TCP_HEADER_OFFSET_DATA_OFFSET_AND_FLAGS);
+        int flags = dataOffsetAndFlags & 0x1F;
+        int dataOffset = (dataOffsetAndFlags & 0xF000) >> 12;
+        int dataLength = packetBuf.readableBytes() - headerOffset - (dataOffset * 4);
+        int checksum = packetBuf.getUnsignedShort(headerOffset + TCP_HEADER_OFFSET_CHECKSUM);
+
+        return new InetUtils.TCPHeader(sourcePort, destPort, sequenceNum, flags, dataLength, checksum);
+    }
+
+    public static final int IP_VERSION_4                        = 0x40;
+    public static final int IP_VERSION_6                        = 0x60;
+
+    public static final int IP_PROTOCOL_TCP                     = 0x06;
+    public static final int IP_PROTOCOL_UDP                     = 0x11;
+    public static final int IP_PROTOCOL_ICMP                    = 0x01;
+    public static final int IP_PROTOCOL_ICMPV6                  = 0x3A;
+
+    public static final int IPV4_MIN_HEADER_LENGTH              = 20;
+    public static final int IPV4_HEADER_OFFSET_VERSION          = 0;
+    public static final int IPV4_HEADER_OFFSET_HEADER_LENGTH    = 0;
+    public static final int IPV4_HEADER_OFFSET_TOTAL_LENGTH     = 2;
+    public static final int IPV4_HEADER_OFFSET_PROTOCOL         = 9;
+    public static final int IPV4_HEADER_OFFSET_CHECKSUM         = 10;
+    public static final int IPV4_HEADER_OFFSET_SOURCE_ADDR      = 12;
+    public static final int IPV4_HEADER_OFFSET_DEST_ADDR        = 16;
+
+    public static final byte[] IPV6_LINK_LOCAL_PREFIX           = { (byte) 0xFE, (byte) 0x80, 0, 0, 0, 0, 0, 0 };
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Make this array "private". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1873)
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Make this member "protected". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS2386)